### ISSUE TITLE
Like at % Watched not Working

### ISSUE THREAD

**Issue by wisnoskij** [Taknok/youtube-auto-like#59]
**Expected Behavior**
Like the video at 50% watched during playback.

**Actual Behavior**
Does not ever like the video unless I refresh the page after reaching the threshold

**To Reproduce**
Steps to reproduce the behaviour:
1. Set the extension to like at any %.
2. Watch video
3. See that it is still unliked
4. Disabled extension and refresh, still unliked
5. Enable extension and refresh
6. See it is liked.

**Specification**
 - OS: Fedora 35
 - 98.0.2
 - V3.4.2

**Comment by Taknok** [comment]
Hi,

Thank you for your feedback ! I just tested in on my Firefox 98 and it is working. Maybe you are in a region were they are testing a new youtube version. Could you send the log and the HTML source of youtube to allow me to replicate on my computer ?
Here is how to: https://www.youtube.com/watch?v=C4fQXRYwbrs&t=66s

**Comment by wisnoskij** [comment]
The dev one says it is the same version, but is working

**Comment by wisnoskij** [comment]
Oh, and now it doesn't work. weird. So I got logs for two different videos one that worked and one that did not. But I assume it is not the video and something external like having the inspection tab open or it is just random. But apparently is is like impossible to zip a file in fedora, so I am not sure how I am going to attach them. But if I just keep installing programs I assume one of them will allow me to zip a archive eventually.

**Comment by wisnoskij** [comment]
[logs.zip](https://github.com/Taknok/youtube-auto-like/files/8403473/logs.zip)

**Comment by Taknok** [comment]
Zip, tar, 7z any standard is okay :)

Your zip is perfect I will look into it to search the origin of the issue. Thank you !

**Comment by Taknok** [comment]
From the logs, it seems that the video is [not detected as a video](https://github.com/Taknok/youtube-auto-like/blob/6c2579b58202cf120c5deb1ac27c0ed8b3a7f374/app/scripts/modules/liker-material.js#L425) (meaning no `watch` [string in the URL](https://github.com/Taknok/youtube-auto-like/blob/6c2579b58202cf120c5deb1ac27c0ed8b3a7f374/app/scripts/modules/liker-material.js#L422)). Could you describe how you start the video ?

Ex:
1. Going to subscription tabs
2. Starting video in small player
3. Pressing `i` to enter in full player
4. Wait to forever to like

**Comment by wisnoskij** [comment]
Oh, That is interesting, and might explain why the only time I noticed it working is the one time I was testing and just searched for a Rick Ashly video on google. Normally I just watch the "watch later" playlist, so maybe the extension does not like playlists? Will test later tonight to see if I can figure out if I can get the ext. to work reliably and fail reliably for me with this new infomation.

**Comment by Taknok** [comment]
The extension totally works with playlist. I'm using also with watch later and other playlists

**Comment by wisnoskij** [comment]
For some reason it seem way harder to get to fail now. But I just encountered one. (this was the only liker log created for the entire video)

> uncaught exception: Multiple run [liker-material.js:440:8](moz-extension://d74800e8-95e3-459b-b36a-74e31b15cd9d/scripts/modules/liker-material.js)


```
await this.update_options();
		this.waitForVideo(() => {
```

It was the only tab open, just watching like normal. Right before that error I got 2 

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://play.google.com/log?format=json&hasfast=true&authuser=0. (Reason: CORS request did not succeed). Status code: (null).

**Comment by Taknok** [comment]
The multiple run exception is normal, normally it doesn't prevent script to like, it is triggered when script is already running, to avoir like > unlike in a raw. It doesn't like the video in your case ?

CORS error is done by youtube and not my script.

**Comment by wisnoskij** [comment]
Its not the video, after that I was able to replicate over and over again until I refreshed the page. Its a first time Youtube is running problem.

After that last problem, every video worked. So I shutdown FF, starting it backup again and the first video shows
[logs3.txt](https://github.com/Taknok/youtube-auto-like/files/8421419/logs3.txt)
So it knows it should like the video, but then even after watched the video many times over and over again, it never actually likes the video. Then every subsequent video shows the already running error and nothing else. Even other tab seems to share this same problem, until I refresh a youtube page and everything works.

The problem also seems to affect the "Windowed Fullscreen" extension, which I saw failing at the same times, I have removed the extension to see if it is perhaps some clash between the two that is causing it. But that along with a tampermoneky autoliking script have been having this problem a month ago back on Chrome/windows so it seems to be OS/browser independent.

**Comment by wisnoskij** [comment]
[logs.zip](https://github.com/Taknok/youtube-auto-like/files/8421525/logs.zip)
Here are the logs and pages I get from 2 random vidoes watched after restarting FF. I never got any autolikes with either of them. It was the same page, so logs2.txt just contains the new logs, and I left all the non liker-mat. stuff because their is not much and it might help.

The only other thing I could try is disabling all extensions but autoliker. But I even turned off my ad block, so it is only the bring back dislike extension that seems to be doing anything at all.

**Comment by wisnoskij** [comment]
OK, so the problem is with [video.duration and video.currentTime](https://github.com/Taknok/youtube-auto-like/blob/6c2579b58202cf120c5deb1ac27c0ed8b3a7f374/app/scripts/modules/liker-material.js#L224). currentTime is showing as 0 and duration is NaN.

**Comment by Taknok** [comment]
Just to keep you updated, I'm trying to reproduce the bug on my computer, but I'm struggling.
ETA 12/04: still trying.

**Comment by wisnoskij** [comment]
Hmm, well I would just consider it Youtubes problem, and maybe they will fix their stuff eventually.

Here is the video-steam object as it appears in my YT.
[console-export-2022-4-12_12-31-31.txt](https://github.com/Taknok/youtube-auto-like/files/8474843/console-export-2022-4-12_12-31-31.txt)

**Comment by wisnoskij** [comment]
So without even logging in, on chrome, in windows 11. I got the exact same issue. So probably some regional YT stuff, but it has been going on for awhile.

**Comment by wisnoskij** [comment]
Lol, ok. now in both chrome and FF, my sidebar is just empty. I dont have subscriptions, i dont have watch later, I just have my home recommended feed. I guess YT just pushed more experimental changes to me? OK, I dont even have my profile menu, so I cannot even report bugs. Ya, I think YT just has me in some alpha software testing program. I pay for premium, so I will have to see if their is anyway to get them to put me back on the stable branch.

**Comment by Taknok** [comment]
Strange indeed. Do you still have the elements missing without any addon activated ?

**Comment by wisnoskij** [comment]
The sidebar came back, but the profile top right menu is still gone. Did a restart into troubleshooting mode and it is the same. I think I said yes to some "do you want to try out some new features" months ago and I assume that is when it all started.

**Comment by wisnoskij** [comment]
Ya, it is just YT being stupid https://www.reddit.com/r/youtube/comments/u26qz3/cleaned_cookies_and_did_a_hard_reset_on_the/

**Comment by Taknok** [comment]
Youtube released a new layout, is this issue still present ?

**Comment by wisnoskij** [comment]
Hmm, Did a reboot and it worked right off, so it seems to be working now. It was still happening a day or two ago for sure, so thus new YT release seemed to have done the trick.

**Comment by Taknok** [comment]
Ok , I keep the issue open then, if within a week there is no issue can you mention it so I can close this issue ?

**Comment by wisnoskij** [comment]
Good call, not sure if difference between restart and shutdown, but this morning it is not working again.

**Comment by RubenKelevra** [comment]
I can confirm that it has some trouble. Maybe some other add-on is causing interference?

I'm using 'Enhancer for YouTube' and 'SponsorBlock' as well.

Enhancer for YouTube is configured to:
- 'Prevent videos loaded in background tabs from automatically playing'
- Pause videos playing in background tabs when a video starts playing in the foreground tab
- Launch a mini player when scrolling down to read comments
- Automatically hide info cards and end screens
- Automatically expand the video player
- Use the available space based on the viewport dimensions to expand the video player
- Hide chat

SponsorBlock is configured to skip everything.

System: Arch/KDE/Wayland with Firefox 100.0.1

**Comment by Taknok** [comment]
The bug only happens in like % or other mode too ?

**Comment by Taknok** [comment]
> I can confirm that it has some trouble. Maybe some other add-on is causing interference?
> 
> I'm using 'Enhancer for YouTube' and 'SponsorBlock' as well.
> 
> Enhancer for YouTube is configured to:
> 
>     * 'Prevent videos loaded in background tabs from automatically playing'
> 
>     * Pause videos playing in background tabs when a video starts playing in the foreground tab
> 
>     * Launch a mini player when scrolling down to read comments
> 
>     * Automatically hide info cards and end screens
> 
>     * Automatically expand the video player
> 
>     * Use the available space based on the viewport dimensions to expand the video player
> 
>     * Hide chat
> 
> 
> SponsorBlock is configured to skip everything.
> 
> System: Arch/KDE/Wayland with Firefox 100.0.1

Those parameters are working for me ... Do you have YouTube premium ? If not this may be A/B testing, can you tell me in which country you are ? So I can use a proxy to see the same YouTube version. Plus, logs and HTML may be helping me to solve the issue quicker: https://www.youtube.com/watch?v=C4fQXRYwbrs&t=66s

I'm asking a lot sorry :/

**Comment by RubenKelevra** [comment]
- Do you have YouTube premium? **Yes**
 - can you tell me in which country you are? **Germany**
 
Debug:

```
yt-autolike start liker-material.js:431:6
options updated liker-material.js:27:6
Get Video. liker-material.js:175:7
No visible ActionsElements found. liker-material.js:41:7
wait 1s for box liker-material.js:158:7
uncaught exception: Multiple run liker-material.js:438:8
ActionsElements found. 2 liker-material.js:57:7
got buttons liker-material.js:124:7
ActionsElements found. liker-material.js:57:7
got buttons liker-material.js:124:7
checking if video is rated liker-material.js:290:6
All mode liker-material.js:331:7
Use list: false liker-material.js:335:6
Should like: true liker-material.js:367:7
```

**Comment by Taknok** [comment]
Seems like the video is found first, the HTML change and video element is not valid anymore, leading to :

> OK, so the problem is with [video.duration and video.currentTime](https://github.com/Taknok/youtube-auto-like/blob/6c2579b58202cf120c5deb1ac27c0ed8b3a7f374/app/scripts/modules/liker-material.js#L224). currentTime is showing as 0 and duration is NaN.
@wisnoskij

**Comment by Taknok** [comment]
When you open the youtube page, the video start playing ASAP ? Or you have to press a play button with a background picture on video frame to start loading and playing the video?

Do you see any HTML change, like switch from normal to theater ?

**Comment by RubenKelevra** [comment]
Not really, it looks like it's in theater right from the moment I load the page. So yeah, it'll switch to theater because of "Enhancer for Youtube".

**Comment by RubenKelevra** [comment]
"Instant like" do work for me fine. I switched to that in the meantime. Would just be nice to "like" videos only when I really watched them for a while.

**Comment by wisnoskij** [comment]
Ya, instant like always worked for me, the only problem seems to be the the script has no way of knowing the current time, everything else works.

**Comment by Taknok** [comment]
> Would just be nice to "like" videos only when I really watched them for a while.

Yes, I'm using % too for this reason. And many other users too (I suppose).

Instant like is working because it does not care about the video element modification.
Here is the simplified instant like workflow:
```
1. Wait the video element
2. Wait the buttons elements
3. Check if liked/disliked on button elements
4. Like
```

The simplified % workflow:
```
1. Wait for video
2. Wait for buttons elements
3. Check if already liked/disliked on buttons elements
4. Wait reaching % using the previous detected video element (to determine current time and duration) {1}
5. Check if not liked by user in the mean time with button elements
6. Like
```
In the % it seems that the video element is modified while waiting {1} so the previously registered video element cannot indicate anymore a proper current time and duration doing an infinite wait.

"My issue" is that I cannot reproduce your bug in order to develop and test a patch. I will try to dev something and make you test it :/ but it is not efficient at all.

**Comment by RubenKelevra** [comment]
@wisnoskij what plugins in Firefox do you use?

@Taknok 

Thanks for the efforts! :)

**Comment by wisnoskij** [comment]
@Taknok If you just want to send some YT API commands, I can try to run them in a console with the bug active to see if they return correct results or fix the issue.

Alternatively, could you just do something like, if (duration = NaN) refresh page;

@RubenKelevra I have tested this issue on multiple OSes, using multiple accounts, on different browsers, and with extensions and without them. It's not an extension issue, the YT API is literally just broken. The only thing I have noticed is that "YouTube Windowed FullScreen" has the exact same issue and wont appear as a in player button if the auto liker is not working.

**Comment by RubenKelevra** [comment]
Hm maybe peek at the Sponsorblock codebase? 🤔 I mean it's working and has to fetch the timecode as well :)

**Comment by Taknok** [comment]
From sponsor source code, it seems that indeed youtube has introduce changes of video element for others elements while playing.

**Comment by Taknok** [comment]
I just pushed something to dev branch to try to fix this issue: https://github.com/Taknok/youtube-auto-like/archive/refs/heads/dev.zip. Can you try it with debug mode (deactivate the extension installed from the store) and tell me if it fix the issue or not (with debug logs if not working please) ?
How to install: https://www.youtube.com/watch?v=C4fQXRYwbrs

**Comment by wisnoskij** [comment]
Hmm, the only time I could get the issue to reproduce itself was when I did not load the extension. So either I am just unlucky and the issue is getting less common, or simply having it loaded fixes other extensions suffering from the same issue as well.

**Comment by Taknok** [comment]
No one can test it ? :/

**Comment by wisnoskij** [comment]
tested again, still working for me.

**Comment by Taknok** [comment]
Tested with the dev version or the store version ?

**Comment by wisnoskij** [comment]
dev, store shows some behaviour

**Comment by Taknok** [comment]
95% with which plugin version ?

**Comment by Taknok** [comment]
Is the issue still present with the version 3.5.0 ?

**Comment by wisnoskij** [comment]
Will verify tomorrow after a reboot. I thought I saw the issue still persist after I saw the update.

**Comment by wisnoskij** [comment]
I have not noticed any problems since the update, but I have been very busy so have not done much explicit testing.

**Comment by wisnoskij** [comment]
I got the issue to happen. If I remember correctly, it used to be reproducible with 100% accuracy after I figured out the trigger. It seems to be like 10-30% chance now, unless it is some issue to do with half loading the home page or some weird trick like that, I cannot reproduce the issue most times, but I just got it to happen twice. Maybe it is based on the order the mods are loaded in or something like that.

**Comment by RubenKelevra** [comment]
> 95% with which plugin version ?

I'm running right now the 3.5.0 The issue for the last 5% is SponsorBlock I think.

Just an educated guess, but I think this is going on:

- Auto Like waits for the time marked for x % to pass, which never happens, because
- Sponsor Block finds a advertisement and skips over this timemarker


The solution would be (in general for all types of skipping) to just count the watch time in the current tab, and if it's over x % a like will be sent.

This also avoids liking content where you just skip to a certain portion and doesn't watch the whole thing. :)

**Comment by Taknok** [comment]
@RubenKelevra The code is already doing this, meaning if you start a vidéo and immediately click to the end, the plugin should like the video. https://github.com/Taknok/youtube-auto-like/blob/e9043abc158fadf61c3f59bda89aa54732cf93a4/app/scripts/modules/liker-material.js#L235

The plugin get the current time of the vidso, meaning if you jump to 30min after 5s of playing the currentTime will be 30min.

I will look the behavior with sponsor block again and on a longer period, because I'm not using it personally.

**Comment by RubenKelevra** [comment]
Ah okay. No I don't think you have to worry about it then. It's just, that sometimes videos don't get liked and I just don't know why.

Regarding your current approach: I think it's better get the actual time the user plays back the video and only after x% of watch time compared to the length of the video it will get liked.

Meaning that if I for example open a video which is clickbait and I skip forward to the highlight portion of the video, it will be liked, despite that I only watched a couple of seconds of it.

Regarding Sponsor Block: Give the add-on a look, its great :)

**Comment by Taknok** [comment]
Even with sponsor block enabled to skip everything, I cannot reproduce the issue with more than 200 videos played :/

**Comment by Taknok** [comment]
If someone having this issue can look into the html code and the addon code. It will gladly help :/

**Comment by Taknok** [comment]
Is the issue fix with the new version of the addon ? (I will close in 7 days without answer, feel free to comment for re-opening it)

**Comment by wisnoskij** [comment]
I forgot to check this morning, but I have seen this issue since the new YT format and the extension update for it. I will try to remember to check specifically the next time I start my computer. The issue seemed unchanged.

**Comment by wisnoskij** [comment]
Turns out my first tab still had this issue ongoing, so I can absolutely state that this issue persists unchanged.

**Comment by Taknok** [comment]
Does 3.6.2 fix this issue ?

**Comment by wisnoskij** [comment]
Yes, I am fairly certain it is working now.

**Comment by Taknok** [comment]
Ok, I will wait one more week before closing it.

**Comment by Taknok** [comment]
Hi,
Can you activate the logs (guide https://youtu.be/C4fQXRYwbrs?t=68) and send it back here when you see the issue ? You just have to activate the log, use youtube normally, and when seeing the bug do `ctrl + shift + K` scroll to find the logs and paste here.

**Comment by wisnoskij** [comment]
Can confirm, just noticed a video not get liked 5 minutes ago. Enabled debug so if it happened again I will have logs.

**Comment by wisnoskij** [comment]
Hmm, yes seems different this time. The videos just don't ever get liked. I even tried instant like and it failed.

https://pastebin.com/CywpPUrj

Looks like they changed the button ids or something like that.

**Comment by Taknok** [comment]
Hi,
I pushed a fix. Can you try the version on dev branch ?  And tell me if the issue is fixed ?
Guide: https://youtu.be/C4fQXRYwbrs?t=68

**Comment by wisnoskij** [comment]
Yes, I tested instant like for every video and it worked right off.

**Comment by Taknok** [comment]
The new version is sent to firefox for validation.

**Comment by Taknok** [comment]
Does the new version fix the issue ?

**Comment by wisnoskij** [comment]
Works for me.

**Comment by Taknok** [comment]
Ok, I will close it in 1 week if no one says otherwise.

**Comment by JoelleEmmily** [comment]
Same issue

**Comment by Taknok** [comment]
Ok, it seems that a new youtube version is released as the script cannot grab the creator block. I will look into it.

**Comment by Taknok** [comment]
Is this issue still present on the new youtube grid version ?

**Comment by Taknok** [comment]
Close as no activity and youtube grid is deprecated
